### PR TITLE
Remove obsolete duplicate APIs from type symbols.

### PR DESCRIPTION
**Pull request by AlekseyTs**
This change removes the following APIs:
- TypeSymbol.BaseType
- TypeSymbol.Interfaces
- TypeSymbol.AllInterfaces
- NamedTypeSymbol.TypeArguments
- TypeParameterSymbol.ConstraintTypes

These APIs simply delegate the work to APIs with "NoUseSiteDiagnostics" suffix and are not supposed to be used inside compilers. They were left in the code-base because they used to be part of the public surface and the names were nicer, but they are no longer part of the public surface.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review this code clean-up change. Majority of changes are fix-ups in C# tests.

File: src/Compilers/Test/Utilities/CSharp/Extensions.cs [hunk 1]
```diff
@@ -554,4 +554,29 @@ internal static Symbol GetDeclaredSymbolFromSyntaxNode(this CSharpSemanticModel
 
         return (Symbol)model.GetDeclaredSymbol(declaration, cancellationToken);
     }
+
+    public static NamedTypeSymbol BaseType(this TypeSymbol symbol)
```
Comment (jcouv) [inline src/Compilers/Test/Utilities/CSharp/Extensions.cs]: I'd be tempted to remove those extensions and update the tests to call the new property/method names instead. But it's not essential.

**Comment by jcouv** [comment]
>         internal override ImmutableArray<NamedTypeSymbol> InterfacesNoUseSiteDiagnostics(ConsList<Symbol> basesBeingResolved)

Should the parameter be optional, to match?

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:671 in 1208772. [](commit_id = 120877284e0d12b59bbe1b7766ac2aae4987fc87, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         internal override ImmutableArray<NamedTypeSymbol> InterfacesNoUseSiteDiagnostics(ConsList<Symbol> basesBeingResolved)

> Should the parameter be optional, to match?

In C# it doesn't have to. 

---
In reply to: [353432669](https://github.com/dotnet/roslyn/pull/23888#issuecomment-353432669) [](ancestors = 353432669)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:671 in 1208772. [](commit_id = 120877284e0d12b59bbe1b7766ac2aae4987fc87, deletion_comment = False)

File: src/EditorFeatures/CSharpTest/SymbolId/SymbolKeyMetadataVsSourceTests.cs [hunk 2]
```diff
@@ -70,7 +70,7 @@ public class App : C
             var member02 = (typesym.GetMembers("Prop").Single() as PropertySymbol).Type;
 
             // 'C'
-            var member03 = typesym.BaseType;
+            var member03 = typesym.BaseType();
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/SymbolId/SymbolKeyMetadataVsSourceTests.cs]: Hmm, these tests really shouldn't be using internals like this; can you file a bug against IDE to get this fixed up?

Comment (AlekseyTs) [reply src/EditorFeatures/CSharpTest/SymbolId/SymbolKeyMetadataVsSourceTests.cs]: I don't think it matters how this test gets to the symbol it needs. It looks like the interesting part is done in ResolveAndVerifySymbol, which operates through public surface.

**Comment by AlekseyTs** [comment]
CC @jaredpar For ask mode approval.

**Comment by jaredpar** [comment]
Approved

**Pull request merged**